clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -467,148 +508,175 @@ class MarkDeviceFunction : public RecursiveASTVisitor<MarkDeviceFunction> {
+    if (llvm::find(CallStack, CurrentDecl) != CallStack.end()) {
+      RecursiveFunctions.insert(CurrentDecl->getCanonicalDecl());
+      return;
+    }
 
-      CallGraphNode *N = SYCLCG.getNode(FD);
-      if (!N)
-        continue;
+    // We previously thought we could skip this function if we'd seen it before,
+    // but if we haven't seen it before in this call graph, we can end up
+    // missing a recursive call.  SO, we have to re-visit call-graphs we've
+    // already seen, just in case it ALSO has recursion.  For example:
+    // void recurse1();
+    // void recurse2() { recurse1(); }
+    // void recurse1() { recurse2(); }
+    // void CallerInKernel() { recurse1(); recurse2(); }
+    // When checking 'recurse1', we'd have ended up 'visiting' recurse2 without
+    // realizing it was recursive, since we enver went into the
+    // child-of-its-child, since THAT was recursive and exited early out of
+    // necessity.
+    // Then when we go to visit the kernel's call to recurse2, we would
+    // immediately escape not noticing it was recursive. SO, we have to do a
+    // little extra work in this case, and make sure we visit the entire call
+    // graph.
+    DeviceFunctions.insert(CurrentDecl);
+
+    // Collect attributes for functions that aren't the root kernel.
+    if (!CallStack.empty()) {
+      bool DirectlyCalled = CallStack.size() == 1;
+      collectSYCLAttributes(Parent.SemaRef, CurrentDecl, CollectedAttributes,
+                            DirectlyCalled);
+    }
 
-      for (const CallGraphNode *CI : *N) {
-        if (auto *Callee = dyn_cast<FunctionDecl>(CI->getDecl())) {
-          Callee = Callee->getMostRecentDecl();
-          if (!Visited.count(Callee))
-            WorkList.push_back({Callee, FD});
-        }
+    // Calculate the kernel body.  Note the 'isSYCLKernelBodyFunction' only
+    // tests that it is operator(), so hopefully this doesn't get us too many
+    // false-positives.
+    if (isSYCLKernelBodyFunction(CurrentDecl)) {
+      // This is a direct callee of the kernel.
+      if (CallStack.size() == 1) {
+        assert(!KernelBody && "inconsistent call graph - only one kernel body "
+                              "function can be called");
+        KernelBody = CurrentDecl;
+      } else if (CallStack.size() == 2 && KernelBody == CallStack.back()) {
+        // To implement rounding-up of a parallel-for range the
+        // SYCL header implementation modifies the kernel call like this:
+        // auto Wrapper = [=](TransformedArgType Arg) {
+        //  if (Arg[0] >= NumWorkItems[0])
+        //    return;
+        //  Arg.set_allowed_range(NumWorkItems);
+        //  KernelFunc(Arg);
+        // };
+        //
+        // This transformation leads to a condition where a kernel body
+        // function becomes callable from a new kernel body function.
+        // Hence this test.
+        // FIXME: We need to be more selective here, this can be hit by simply
```

Comment:
Not sure I'm going to fix this HERE, but we probably SHOULD eventually. I think we need to add an attribute that means "this is a kernel wrapper" that the library uses to highlight these situations (and likely permit wrappers of wrappers if necessary!).

Reply:
Thanks, this will be helpful!